### PR TITLE
GH-109214: `_SET_IP` before `_PUSH_FRAME` (but not `_POP_FRAME`)

### PR DESCRIPTION
**Pull request by brandtbucher**
#110755 inadvertently allowed the tier two optimizer to remove instruction pointer updates before frame pushes, which breaks the following return. This updates `remove_unneeded_uops` to recognize `_PUSH_FRAME` as a uop that requires an up-to-date instruction pointer on the frame.

While investigating this, I realized that we actually *don't* need to set the instruction pointer before returning from a frame (since the frame is unlinked and torn down immediately after). So I've also removed `_SAVE_CURRENT_IP` from both of the return opcodes, which is a nice little win for tier one.

Also, remove some dead code and comment/assert a non-obvious bit of code during trace projection.

* Issue: gh-109214

### CODE REVIEW

File: Python/optimizer.c
Comment (gvanrossum) [inline Python/optimizer.c:684]: If this is always it, why do we need to smuggle it in via offset? Also it's beginning to look like the offset should just be made one larger in the code generator, so we can drop the ugly `+ 1` on line 654 above.

**Review by gvanrossum — APPROVED** [review]
I can't say I am *less* happy with this than with the situation before the bug was introduced, but I am also not sure I am *happier*. It feels as if the various special uops SET_IP, SAVE_CURRENT_IP, POP/PUSH_FRAME can still use more work (not in this PR though).

**Comment by brandtbucher** [comment]
I agree with your comments. I'll land this now to unblock tier two, and work on a follow-up PR to clean up this IP stuff further.

**Comment by gvanrossum** [comment]
> I agree with your comments. I'll land this now to unblock tier two, and work on a follow-up PR to clean up this IP stuff further.

Maybe wait until Irit's PR lands. It touches on all these things.

**Comment by brandtbucher** [comment]
Okay, I will.

**Pull request merged**
